### PR TITLE
CSEN-25 DAO classes were implemented

### PR DESCRIPTION
**Pull request by MedvedevBW**

### CODE REVIEW

**Comment by MedvedevBW** [comment]
@placccebo

File: src/main/scala/com/bwsw/cloudstack/KeyAuthenticationClientCreator.scala [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+* under the License.
+*/
+package com.bwsw.cloudstack
+
+import br.com.autonomiccs.apacheCloudStack.client.ApacheCloudStackClient
+import br.com.autonomiccs.apacheCloudStack.client.beans.ApacheCloudStackUser
+import org.slf4j.LoggerFactory
+
+/**
+  * Class is responsible for ApacheCloudStackClient creation
```
Comment (placccebo) [inline src/main/scala/com/bwsw/cloudstack/KeyAuthenticationClientCreator.scala]: It's a very general description and two classes have the same description. Add some specific info

File: src/main/scala/com/bwsw/cloudstack/KeyAuthenticationClientCreator.scala [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+  override def createClient(endpoint: String): ApacheCloudStackClient = {
+    logger.trace(s"createClient(endpoint: $endpoint)")
+    new ApacheCloudStackClient(endpoint, apacheCloudStackUser)
+  }
+}
+
+object KeyAuthenticationClientCreator {
+  /**
+    * Class is responsible for providing settings for ApacheCloudStackClient creation
```
Comment (placccebo) [inline src/main/scala/com/bwsw/cloudstack/KeyAuthenticationClientCreator.scala]: the same comment

File: src/main/scala/com/bwsw/cloudstack/entities/dao/AccountDao.scala [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+import br.com.autonomiccs.apacheCloudStack.exceptions.ApacheCloudStackClientRequestRuntimeException
+import com.bwsw.cloudstack.entities.Executor
+import com.bwsw.cloudstack.entities.common.JsonMapper
+import com.bwsw.cloudstack.entities.requests.account.{AccountCreateRequest, AccountFindRequest}
+import com.bwsw.cloudstack.entities.responses.{Account, AccountResponse}
+
+import scala.util.{Failure, Success, Try}
+
+/**
+  * Class is responsible for creating and reading ApacheCloudStack accounts
```
Comment (placccebo) [inline src/main/scala/com/bwsw/cloudstack/entities/dao/AccountDao.scala]: reading -> retrieving

File: src/main/scala/com/bwsw/cloudstack/entities/dao/AccountDao.scala [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    * @param request see: [[AccountCreateRequest]]
+    */
+  override def create(request: C): Unit = {
+    Try {
+      super.create(request)
+    } match {
+      case Success(x) =>
+        logger.debug(s"Account was created by request: $request")
+      case Failure(e: Throwable) =>
+        logger.error(s"Can not to create account, exception: $e was thrown")
```
Comment (placccebo) [inline src/main/scala/com/bwsw/cloudstack/entities/dao/AccountDao.scala]: Can't create an account  by request: $request

File: src/main/scala/com/bwsw/cloudstack/entities/dao/AccountDao.scala [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  /**
+    * Retrieve accounts
+    *
+    * @param request see: [[AccountFindRequest]]
+    */
+  override def find(request: F): List[Account] = {
+    logger.trace(s"find(request: $request)")
+    val accounts = Try {
+      val response = executor.executeRequest(request.request)
+      mapper.deserialize[AccountResponse](response).entityList.entities.getOrElse(List.empty[Account])
```
Comment (placccebo) [inline src/main/scala/com/bwsw/cloudstack/entities/dao/AccountDao.scala]: AccountResponse should be a type parameter

File: src/main/scala/com/bwsw/cloudstack/entities/dao/AccountDao.scala [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    } match {
+      case Success(x) =>
+        logger.debug(s"Accounts were retrieved: $x")
+        x
+      case Failure(e: ApacheCloudStackClientRequestRuntimeException) if e.getStatusCode == ENTITY_DOES_NOT_EXIST =>
+        logger.warn(s"No accounts found on request: $request")
```
Comment (placccebo) [inline src/main/scala/com/bwsw/cloudstack/entities/dao/AccountDao.scala]: were found

File: src/main/scala/com/bwsw/cloudstack/entities/dao/AccountDao.scala [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        List.empty[Account]
+      case Failure(e: Throwable) =>
+        logger.error(s"Can not to find accounts, an exception: $e was thrown")
```
Comment (placccebo) [inline src/main/scala/com/bwsw/cloudstack/entities/dao/AccountDao.scala]: Can't find any account by request: $request, exception: $e was thrown

File: src/test/scala/com/bwsw/cloudstack/KeyAuthenticationClientCreatorTestSuite.scala [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+import br.com.autonomiccs.apacheCloudStack.client.ApacheCloudStackClient
+import br.com.autonomiccs.apacheCloudStack.client.beans.ApacheCloudStackUser
+import com.bwsw.cloudstack.entities.Executor
+import com.bwsw.cloudstack.entities.common.JsonMapper
+import com.bwsw.cloudstack.entities.dao.UserDao
+import com.bwsw.cloudstack.entities.requests.user.{UserCreateRequest, UserFindRequest}
+import org.scalatest.FlatSpec
+
+class KeyAuthenticationClientCreatorTestSuite extends FlatSpec {
+  "createClient" should "create ApacheCloudStackClient which include ApacheCloudStackUser authenticated by keys" in {
```
Comment (placccebo) [inline src/test/scala/com/bwsw/cloudstack/KeyAuthenticationClientCreatorTestSuite.scala]: include -> contains

File: src/test/scala/com/bwsw/cloudstack/PasswordAuthenticationClientCreatorTestSuite.scala [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+*/
+package com.bwsw.cloudstack
+
+import br.com.autonomiccs.apacheCloudStack.client.ApacheCloudStackClient
+import br.com.autonomiccs.apacheCloudStack.client.beans.ApacheCloudStackUser
+import org.scalatest.FlatSpec
+
+class PasswordAuthenticationClientCreatorTestSuite extends FlatSpec {
+
+  "createClient" should "create ApacheCloudStackClient which include ApacheCloudStackUser authenticated by keys" in {
```
Comment (placccebo) [inline src/test/scala/com/bwsw/cloudstack/PasswordAuthenticationClientCreatorTestSuite.scala]: fix the description

File: src/test/scala/com/bwsw/cloudstack/entities/dao/AccountDaoTestSuite.scala [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+import br.com.autonomiccs.apacheCloudStack.exceptions.ApacheCloudStackClientRequestRuntimeException
+import com.bwsw.cloudstack.entities.Executor
+import com.bwsw.cloudstack.entities.requests.account.{AccountCreateRequest, AccountFindRequest}
+import com.bwsw.cloudstack.entities.responses.{Account, User}
+import org.scalatest.FlatSpec
+
+class AccountDaoTestSuite extends FlatSpec with TestData {
+  val findRequest = new AccountFindRequest
+
+  "find" should "return an entity list if response json string contains data" in {
```
Comment (placccebo) [inline src/test/scala/com/bwsw/cloudstack/entities/dao/AccountDaoTestSuite.scala]: "find" should "return non-empty entity list if a response json string contains the relevant data"

File: src/test/scala/com/bwsw/cloudstack/entities/dao/AccountDaoTestSuite.scala [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+        Response.getAccountResponseJson(accountId.toString, userId.toString)
+      }
+    }
+
+    val accountDao = new AccountDao(executor, jsonMapper)
+
+    assert(accountDao.find(findRequest) == expectedAccountList)
+  }
+
+  "find" should "return an empty entity list if response json string does not contain data" in {
```
Comment (placccebo) [inline src/test/scala/com/bwsw/cloudstack/entities/dao/AccountDaoTestSuite.scala]: add an article and "relevant" word

File: src/test/scala/com/bwsw/cloudstack/entities/dao/AccountDaoTestSuite.scala [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+        throw new Exception
+      }
+    }
+
+    val accountDao = new AccountDao(executor, jsonMapper)
+
+    assertThrows[Exception](accountDao.find(findRequest))
+  }
+
+  "find" should "not swallow ApacheCloudStackClientRequestRuntimeException with status other than 431" in {
```
Comment (placccebo) [inline src/test/scala/com/bwsw/cloudstack/entities/dao/AccountDaoTestSuite.scala]: with a status different from 431

**Review by placccebo — CHANGES_REQUESTED** [review]

File: src/main/scala/com/bwsw/cloudstack/KeyAuthenticationClientCreator.scala (hunk 1 above)
Comment (MedvedevBW) [reply src/main/scala/com/bwsw/cloudstack/KeyAuthenticationClientCreator.scala]: ok

File: src/main/scala/com/bwsw/cloudstack/KeyAuthenticationClientCreator.scala (hunk 2 above)
Comment (MedvedevBW) [reply src/main/scala/com/bwsw/cloudstack/KeyAuthenticationClientCreator.scala]: ok

File: src/main/scala/com/bwsw/cloudstack/entities/dao/AccountDao.scala (hunk 3 above)
Comment (MedvedevBW) [reply src/main/scala/com/bwsw/cloudstack/entities/dao/AccountDao.scala]: ok

File: src/main/scala/com/bwsw/cloudstack/entities/dao/AccountDao.scala (hunk 4 above)
Comment (MedvedevBW) [reply src/main/scala/com/bwsw/cloudstack/entities/dao/AccountDao.scala]: request for account creation include login and password, so I think that to show it in "error" logs is unsafe

File: src/main/scala/com/bwsw/cloudstack/entities/dao/AccountDao.scala (hunk 5 above)
Comment (MedvedevBW) [reply src/main/scala/com/bwsw/cloudstack/entities/dao/AccountDao.scala]: AccountResponse does not present in method signature and it's been using only in internal logic, so I think adding type parameter for it is unnecessarily

File: src/main/scala/com/bwsw/cloudstack/entities/dao/AccountDao.scala (hunk 6 above)
Comment (MedvedevBW) [reply src/main/scala/com/bwsw/cloudstack/entities/dao/AccountDao.scala]: ok

File: src/main/scala/com/bwsw/cloudstack/entities/dao/AccountDao.scala (hunk 7 above)
Comment (MedvedevBW) [reply src/main/scala/com/bwsw/cloudstack/entities/dao/AccountDao.scala]: account or accounts? I used "accounts"

File: src/test/scala/com/bwsw/cloudstack/KeyAuthenticationClientCreatorTestSuite.scala (hunk 8 above)
Comment (MedvedevBW) [reply src/test/scala/com/bwsw/cloudstack/KeyAuthenticationClientCreatorTestSuite.scala]: ok

File: src/test/scala/com/bwsw/cloudstack/PasswordAuthenticationClientCreatorTestSuite.scala (hunk 9 above)
Comment (MedvedevBW) [reply src/test/scala/com/bwsw/cloudstack/PasswordAuthenticationClientCreatorTestSuite.scala]: ok

File: src/test/scala/com/bwsw/cloudstack/entities/dao/AccountDaoTestSuite.scala (hunk 10 above)
Comment (MedvedevBW) [reply src/test/scala/com/bwsw/cloudstack/entities/dao/AccountDaoTestSuite.scala]: ok

File: src/test/scala/com/bwsw/cloudstack/entities/dao/AccountDaoTestSuite.scala (hunk 11 above)
Comment (MedvedevBW) [reply src/test/scala/com/bwsw/cloudstack/entities/dao/AccountDaoTestSuite.scala]: ok

File: src/test/scala/com/bwsw/cloudstack/entities/dao/AccountDaoTestSuite.scala (hunk 12 above)
Comment (MedvedevBW) [reply src/test/scala/com/bwsw/cloudstack/entities/dao/AccountDaoTestSuite.scala]: ok

**Comment by MedvedevBW** [comment]
@placccebo

File: src/main/scala/com/bwsw/cloudstack/KeyAuthenticationClientCreator.scala [hunk 13]
```diff
@@ -39,7 +39,7 @@ class KeyAuthenticationClientCreator(settings: KeyAuthenticationClientCreator.Se
 
 object KeyAuthenticationClientCreator {
   /**
-    * Class is responsible for providing settings for ApacheCloudStackClient creation
+    * Class is responsible for providing authentication keys settings for ApacheCloudStackClient creation
```
Comment (placccebo) [inline src/main/scala/com/bwsw/cloudstack/KeyAuthenticationClientCreator.scala]: you should use only one word: either keys or settings

File: src/main/scala/com/bwsw/cloudstack/PasswordAuthenticationClientCreator.scala [hunk 14]
```diff
@@ -39,7 +39,7 @@ class PasswordAuthenticationClientCreator(settings: PasswordAuthenticationClient
 
 object PasswordAuthenticationClientCreator {
   /**
-    * Class is responsible for providing settings for ApacheCloudStackClient creation
+    * Class is responsible for providing 'login/password' authentication settings for ApacheCloudStackClient creation
```
Comment (placccebo) [inline src/main/scala/com/bwsw/cloudstack/PasswordAuthenticationClientCreator.scala]: Class is responsible for providing a login and a password for ApacheCloudStackClient creation

File: src/main/scala/com/bwsw/cloudstack/entities/dao/AccountDao.scala (hunk 4 above)
Comment (placccebo) [reply src/main/scala/com/bwsw/cloudstack/entities/dao/AccountDao.scala]: ok, remove a request from the log

**Review by placccebo — CHANGES_REQUESTED** [review]

File: src/main/scala/com/bwsw/cloudstack/KeyAuthenticationClientCreator.scala (hunk 13 above)
Comment (MedvedevBW) [reply src/main/scala/com/bwsw/cloudstack/KeyAuthenticationClientCreator.scala]: ok

File: src/main/scala/com/bwsw/cloudstack/PasswordAuthenticationClientCreator.scala (hunk 14 above)
Comment (MedvedevBW) [reply src/main/scala/com/bwsw/cloudstack/PasswordAuthenticationClientCreator.scala]: ok

**Comment by MedvedevBW** [comment]
@placccebo

**Review by placccebo — APPROVED** [review]

**Pull request merged**